File: api-test/test-api.py
#!/usr/bin/env python3
"""
Python test script for Relay's file API (/api/files/*)

Usage:
    python test-api.py YOUR_API_KEY [BASE_URL]

Example:
    python test-api.py vbc_abc123xyz http://localhost:3000
"""

import sys
import time

try:
    import requests
except ImportError:
    print("Error: requests library not installed")
    print("Install it with: pip install requests")
    sys.exit(1)


class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    BLUE = '\033[94m'
    YELLOW = '\033[93m'
    END = '\033[0m'


def print_colored(text, color):
    print(f"{color}{text}{Colors.END}")


def print_success(text):
    print_colored(f"✓ {text}", Colors.GREEN)


def print_error(text):
    print_colored(f"✗ {text}", Colors.RED)


def print_info(text):
    print_colored(f"ℹ {text}", Colors.BLUE)


def print_warning(text):
    print_colored(f"⚠ {text}", Colors.YELLOW)


def print_header(text):
    print()
    print_colored("=" * 60, Colors.BLUE)
    print_colored(text, Colors.BLUE)
    print_colored("=" * 60, Colors.BLUE)
    print()


class RelayTester:
    def __init__(self, api_key, base_url):
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.api_url = f"{self.base_url}/api/files"
        self.session = requests.Session()
        self.session.headers.update({'Authorization': f'Bearer {api_key}'})
        self.test_file_id = None
        self.test_short_id = None
        self.test_content = None

    def api_request(self, endpoint, method='GET', **kwargs):
        url = f"{self.api_url}{endpoint}"
        response = self.session.request(method, url, **kwargs)

        try:
            data = response.json()
        except ValueError:
            data = {'error': response.text}

        if not response.ok or data.get('success') is False:
            raise Exception(f"HTTP {response.status_code}: {data.get('error', 'Request failed')}")

        return data

    def test_upload(self):
        print_header("Test 1: Upload File")

        self.test_content = f"Hello from Python test at {time.strftime('%Y-%m-%d %H:%M:%S')}".encode('utf-8')
        filename = f"test-{int(time.time())}.txt"

        print_info(f"Uploading test file: {filename} ({len(self.test_content)} bytes)")

        result = self.api_request('/upload', method='POST', files={
            'file': (filename, self.test_content, 'text/plain'),
        })

        self.test_file_id = result['data']['id']
        self.test_short_id = result['data']['shortId']

        print_success("File uploaded successfully")
        print(f"  ID: {self.test_file_id}")
        print(f"  URL: {result['data']['url']}")

        print_success("Test passed")

    def test_get_file_info(self):
        if not self.test_short_id:
            print_warning("Skipping - no file uploaded yet")
            return

        print_header("Test 2: Get File Info (public lookup)")
        print_info(f"Getting info for file: {self.test_short_id}")

        response = requests.get(f"{self.api_url}/info", params={'file_code': self.test_short_id})
        result = response.json()
        info = result['result'][0]

        if info['status'] != 200:
            raise Exception(f"File info lookup failed: status {info['status']}")

        print("\nFile Info:")
        print(f"  Name: {info['name']}")
        print(f"  Size: {info['size']} bytes")
        print(f"  Uploaded: {info['uploaded']}")

        print_success("Test passed")

    def test_list_files(self):
        print_header("Test 3: List Files")
        print_info("Fetching files list (filtered to text files)...")

        result = self.api_request('/list', params={'limit': 10, 'mimeType': 'text/'})

        print("\nFiles List:")
        print(f"  Total files: {result['pagination']['total']}")
        print(f"  Returned: {len(result['data'])}")

        for i, f in enumerate(result['data'][:3], 1):
            print(f"  {i}. {f['name']} ({f['size']} bytes)")

        print_success("Test passed")

    def test_download(self):
        if not self.test_file_id:
            print_warning("Skipping - no file uploaded yet")
            return

        print_header("Test 4: Download File")

        print_info("Getting download URL...")
        result = self.api_request(f'/download/{self.test_file_id}')
        download_url = result['data']['url']

        print_info("Downloading file...")
        response = requests.get(download_url)

        if not response.ok:
            raise Exception(f"Download failed: HTTP {response.status_code}")

        if response.content == self.test_content:
            print_success("Content matches uploaded data")
        else:
            print_warning("Content does not match uploaded data")

        print_success("Test passed")

    def test_delete(self):
        if not self.test_file_id:
            print_warning("Skipping - no file uploaded yet")
            return

        print_header("Test 5: Delete File")
        print_info(f"Deleting file: {self.test_file_id}")

        try:
            self.api_request(f'/delete', method='DELETE', params={'fileId': self.test_file_id})
            print_success("File deleted successfully")
        except Exception as e:
            if "403" in str(e):
                print_warning("Delete permission not granted (this is expected)")
            else:
                raise

        print_success("Test passed")

    def test_batch_delete(self):
        print_header("Test 6: Batch Delete")

        ids = []
        for i in range(2):
            filename = f"batch-{int(time.time())}-{i}.txt"
            result = self.api_request('/upload', method='POST', files={
                'file': (filename, f"batch test {i}".encode(), 'text/plain'),
            })
            ids.append(result['data']['id'])

        print_info(f"Batch deleting {len(ids)} files...")
        result = self.api_request('/batch-delete', method='POST', json={'fileIds': ids})

        print(f"\nDeleted {result['data']['deletedCount']} of {len(ids)}")
        if result['data']['deletedCount'] != len(ids):
            raise Exception("Not all files were deleted")

        print_success("Test passed")

    def run_all_tests(self):
        print_header("Relay File API Test Suite (Python)")
        print(f"API Key: {self.api_key[:15]}...")
        print(f"Base URL: {self.base_url}")

        tests = [
            ('Upload File', self.test_upload),
            ('Get File Info', self.test_get_file_info),
            ('List Files', self.test_list_files),
            ('Download File', self.test_download),
            ('Delete File', self.test_delete),
            ('Batch Delete', self.test_batch_delete),
        ]

        passed = 0
        failed = 0

        for name, test_fn in tests:
            try:
                test_fn()
                passed += 1
                time.sleep(0.5)
            except Exception as e:
                print_error(f"Test failed: {e}")
                failed += 1

        print_header("Test Summary")
        print(f"Total Tests: {passed + failed}")
        print_success(f"Passed: {passed}")
        if failed > 0:
            print_error(f"Failed: {failed}")

        print_colored("\n" + "=" * 60, Colors.BLUE)

        if failed == 0:
            print_success("\n🎉 All tests passed! Your API is working correctly!")
        else:
            print_warning("\n⚠️  Some tests failed. Check the errors above.")

        print()


def main():
    if len(sys.argv) < 2:
        print("Usage: python test-api.py YOUR_API_KEY [BASE_URL]")
        print("\nExample:")
        print("  python test-api.py vbc_abc123xyz http://localhost:3000")
        sys.exit(1)

    api_key = sys.argv[1]
    base_url = sys.argv[2] if len(sys.argv) > 2 else 'http://localhost:3000'

    tester = RelayTester(api_key, base_url)

    try:
        tester.run_all_tests()
    except KeyboardInterrupt:
        print("\n\nTests interrupted by user")
        sys.exit(1)
    except Exception as e:
        print_error(f"\nFatal error: {e}")
        sys.exit(1)


if __name__ == '__main__':
    main()

File: api-test/test-api.js
/**
 * API Test Script for Relay's file API (/api/files/*)
 *
 * This script tests the main /api/files endpoints end to end.
 *
 * Setup:
 * 1. npm install (Node 18+ has native fetch/FormData/Blob - no extra deps needed)
 * 2. Set your API key and base URL below
 * 3. Run: node test-api.js
 */

// ===== CONFIGURATION =====
const API_KEY = 'YOUR_API_KEY_HERE'; // Replace with your actual API key
const BASE_URL = 'http://localhost:3000'; // Or your deployed URL
// =========================

const API_URL = `${BASE_URL}/api/files`;

const colors = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  reset: '\x1b[0m',
};

function log(message, color = 'reset') {
  console.log(`${colors[color]}${message}${colors.reset}`);
}

function logSuccess(message) { log(`✓ ${message}`, 'green'); }
function logError(message) { log(`✗ ${message}`, 'red'); }
function logInfo(message) { log(`ℹ ${message}`, 'blue'); }
function logWarning(message) { log(`⚠ ${message}`, 'yellow'); }

async function testEndpoint(name, testFn) {
  log(`\n${'='.repeat(60)}`, 'blue');
  log(`Testing: ${name}`, 'blue');
  log('='.repeat(60), 'blue');

  try {
    await testFn();
    logSuccess(`${name} - PASSED`);
    return true;
  } catch (error) {
    logError(`${name} - FAILED`);
    logError(`Error: ${error.message}`);
    if (error.response) console.log('Response:', error.response);
    return false;
  }
}

async function apiRequest(path, options = {}) {
  const response = await fetch(`${API_URL}${path}`, {
    ...options,
    headers: {
      'Authorization': `Bearer ${API_KEY}`,
      ...options.headers,
    },
  });

  const data = await response.json();

  if (!response.ok || data.success === false) {
    const error = new Error(data.error || 'Request failed');
    error.response = data;
    error.status = response.status;
    throw error;
  }

  return data;
}

// Test 1: Upload File
async function testUpload() {
  logInfo('Creating and uploading test file...');

  const testContent = 'Hello from API test! This is a test file created at ' + new Date().toISOString();
  const testFileName = `test-${Date.now()}.txt`;

  const form = new FormData();
  form.append('file', new Blob([testContent], { type: 'text/plain' }), testFileName);

  const result = await apiRequest('/upload', { method: 'POST', body: form });

  console.log('\nUpload Result:');
  console.log('  ID:', result.data.id);
  console.log('  Short ID:', result.data.shortId);
  console.log('  URL:', result.data.url);

  global.testFileId = result.data.id;
  global.testShortId = result.data.shortId;
  global.testContent = testContent;

  logSuccess('File uploaded successfully!');
}

// Test 2: Get File Info (public, by short ID)
async function testGetFileInfo() {
  if (!global.testShortId) throw new Error('No test file uploaded yet.');

  logInfo(`Getting info for file: ${global.testShortId}`);

  const response = await fetch(`${API_URL}/info?file_code=${global.testShortId}`);
  const result = await response.json();
  const info = result.result[0];

  if (info.status !== 200) throw new Error(`File info lookup failed: status ${info.status}`);

  console.log('\nFile Info:');
  console.log('  Name:', info.name);
  console.log('  Size:', info.size, 'bytes');
  console.log('  Uploaded:', info.uploaded);
}

// Test 3: List Files
async function testListFiles() {
  logInfo('Listing files (filtered to text files)...');

  const result = await apiRequest('/list?limit=10&mimeType=text/');

  console.log('\nFiles List:');
  console.log('  Total files:', result.pagination.total);
  console.log('  Returned:', result.data.length);

  result.data.slice(0, 3).forEach((f, i) => {
    console.log(`  ${i + 1}. ${f.name} (${f.size} bytes)`);
  });
}

// Test 4: Download File
async function testDownload() {
  if (!global.testFileId) throw new Error('No test file uploaded yet.');

  logInfo('Getting download URL...');
  const result = await apiRequest(`/download/${global.testFileId}`);

  logInfo('Downloading file...');
  const response = await fetch(result.data.url);
  if (!response.ok) throw new Error(`Download failed with status ${response.status}`);

  const content = await response.text();
  console.log('\nDownloaded Content:');
  console.log('  Length:', content.length, 'bytes');

  if (content === global.testContent) {
    logSuccess('Content matches uploaded data!');
  } else {
    logWarning('Content does not match expected data');
  }
}

// Test 5: Delete File
async function testDelete() {
  if (!global.testFileId) throw new Error('No test file uploaded yet.');

  logInfo(`Deleting file: ${global.testFileId}`);

  try {
    await apiRequest(`/delete?fileId=${global.testFileId}`, { method: 'DELETE' });
    logSuccess('File deleted successfully!');
  } catch (error) {
    if (error.status === 403) {
      logWarning('Delete permission not granted for this API key (expected if you disabled it)');
    } else {
      throw error;
    }
  }
}

// Test 6: Batch Delete
async function testBatchDelete() {
  logInfo('Uploading two throwaway files for batch delete...');

  const ids = [];
  for (let i = 0; i < 2; i++) {
    const form = new FormData();
    form.append('file', new Blob([`batch test ${i}`], { type: 'text/plain' }), `batch-${Date.now()}-${i}.txt`);
    const result = await apiRequest('/upload', { method: 'POST', body: form });
    ids.push(result.data.id);
  }

  logInfo(`Batch deleting ${ids.length} files...`);
  const result = await apiRequest('/batch-delete', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ fileIds: ids }),
  });

  console.log('\nBatch Delete Result:');
  console.log('  Deleted:', result.data.deletedCount, 'of', ids.length);

  if (result.data.deletedCount !== ids.length) {
    throw new Error('Not all files were deleted');
  }
  logSuccess('Batch delete succeeded!');
}

async function runTests() {
  log('\n' + '='.repeat(60), 'blue');
  log('       Relay File API Test Suite', 'blue');
  log('='.repeat(60), 'blue');

  if (API_KEY === 'YOUR_API_KEY_HERE') {
    logError('\n✗ Please set your API_KEY in the script before running tests!');
    logInfo('1. Visit http://localhost:3000/api (or /developers, depending on your deployment)');
    logInfo('2. Create an API key');
    logInfo('3. Replace YOUR_API_KEY_HERE with your actual key');
    process.exit(1);
  }

  console.log(`\nConfiguration:`);
  console.log(`  API Key: ${API_KEY.substring(0, 10)}...`);
  console.log(`  Base URL: ${BASE_URL}`);

  const results = { passed: 0, failed: 0 };

  const tests = [
    ['Upload File', testUpload],
    ['Get File Info', testGetFileInfo],
    ['List Files', testListFiles],
    ['Download File', testDownload],
    ['Delete File', testDelete],
    ['Batch Delete', testBatchDelete],
  ];

  for (const [name, testFn] of tests) {
    const success = await testEndpoint(name, testFn);
    if (success) results.passed++; else results.failed++;
    await new Promise((resolve) => setTimeout(resolve, 500));
  }

  log('\n' + '='.repeat(60), 'blue');
  log('                    Test Summary', 'blue');
  log('='.repeat(60), 'blue');
  console.log(`\nTotal Tests: ${results.passed + results.failed}`);
  logSuccess(`Passed: ${results.passed}`);
  if (results.failed > 0) logError(`Failed: ${results.failed}`);

  log('\n' + '='.repeat(60), 'blue');

  if (results.failed === 0) {
    logSuccess('\n🎉 All tests passed! Your API is working correctly!');
  } else {
    logWarning('\n⚠️  Some tests failed. Check the errors above.');
  }

  console.log('');
}

runTests().catch((error) => {
  logError('\nFatal error running tests:');
  console.error(error);
  process.exit(1);
});
